core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1486,9 +1486,28 @@ public void delete(AlluxioURI path, DeleteOptions options) throws IOException,
         LockedInodePath inodePath =
             mInodeTree.lockInodePath(lockingScheme.getPath(), lockingScheme.getMode());
         FileSystemMasterAuditContext auditContext =
-            createAuditContext("delete", path, null, inodePath.getInodeOrNull())) {
+            createAuditContext("delete", path, null, inodePath.getInodeOrNull());
+        LockedInodePathList children =
+            options.isRecursive() ? mInodeTree.lockDescendants(inodePath, InodeTree.LockMode.WRITE)
```

Comment:
when is the lock of `children` released? perhaps put it in a `try-with-resources` block?

Reply:
@yuzhu it is already in one

Reply:
@madanadit oops, not enough context on github for me to see the try block. My other suggestion is that since we `lockDescendants` here, maybe we can pass the  `children` object to `deleteInternal` so that we can avoid calling the relatively expensive `lockDescendants` twice?